File: gremlin-dotnet/src/Gremlin.Net/Driver/Messages/ResponseMessage.cs
#region License

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#endregion

namespace Gremlin.Net.Driver.Messages
{
    /// <summary>
    ///     The 4.0 response message returned from the server.
    /// </summary>
    /// <typeparam name="T">The type of the result data.</typeparam>
    public record ResponseMessage<T>
    {
        /// <summary>
        ///     Initializes a new instance of the <see cref="ResponseMessage{T}" /> record.
        /// </summary>
        /// <param name="bulked">Whether the response contains bulked results.</param>
        /// <param name="result">The deserialized result data.</param>
        /// <param name="statusCode">The status code from the GraphBinary response status (after the end-of-stream marker).</param>
        /// <param name="statusMessage">Optional status message from the GraphBinary response status.</param>
        /// <param name="exception">Optional exception string from the GraphBinary response status.</param>
        public ResponseMessage(bool bulked, T result, int statusCode,
            string? statusMessage, string? exception)
        {
            Bulked = bulked;
            Result = result;
            StatusCode = statusCode;
            StatusMessage = statusMessage;
            Exception = exception;
        }

        /// <summary>
        ///     Gets whether the response contains bulked results.
        /// </summary>
        public bool Bulked { get; }

        /// <summary>
        ///     Gets the deserialized result data.
        /// </summary>
        public T Result { get; }

        /// <summary>
        ///     Gets the status code from the GraphBinary response status (200 = success).
        /// </summary>
        public int StatusCode { get; }

        /// <summary>
        ///     Gets the optional status message from the GraphBinary response status.
        /// </summary>
        public string? StatusMessage { get; }

        /// <summary>
        ///     Gets the optional exception string from the GraphBinary response status.
        /// </summary>
        public string? Exception { get; }
    }
}

File: gremlin-util/src/main/java/org/apache/tinkerpop/gremlin/util/message/ResponseMessage.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.util.message;

import io.netty.handler.codec.http.HttpResponseStatus;

import java.util.Collections;
import java.util.List;

/**
 * The model for a response message that is sent to the server beginning in 4.0.0. ResponseMessage is designed to be
 * streamed back the client in parts so depending on the state of the transfer, certain parts may be null at different
 * times.
 *
 * @author Stephen Mallette (http://stephen.genoprime.com)
 */
public final class ResponseMessage {
    private final ResponseStatus responseStatus;
    private final ResponseResult responseResult;

    private ResponseMessage(final ResponseStatus responseStatus,
                            final ResponseResult responseResult) {
        this.responseResult = responseResult;
        this.responseStatus = responseStatus;
    }

    public ResponseStatus getStatus() {
        return responseStatus;
    }

    public ResponseResult getResult() {
        return responseResult;
    }

    @Override
    public String toString() {
        return "ResponseMessage{" +
                ", status=" + responseStatus +
                ", result=" + responseResult +
                '}';
    }

    public static class ResponseMessageHeader {
        private final ResponseMessage responseMessage;
        private final boolean typed;

        public ResponseMessageHeader(final ResponseMessage responseMessage, final boolean typed) {
            this.responseMessage = responseMessage;
            this.typed = typed;
        }

        public ResponseMessage getResponseMessage() {
            return responseMessage;
        }

        public boolean getTyped() {
            return typed;
        }
    }

    public static class ResponseMessageFooter {
        private final ResponseMessage responseMessage;
        private final boolean typed;

        public ResponseMessageFooter(final ResponseMessage responseMessage, final boolean typed) {
            this.responseMessage = responseMessage;
            this.typed = typed;
        }

        public ResponseMessage getResponseMessage() {
            return responseMessage;
        }

        public boolean getTyped() {
            return typed;
        }
    }

    public static Builder build() {
        return new Builder();
    }

    public final static class Builder {
        private HttpResponseStatus code = null;
        private List<Object> result = Collections.emptyList();
        private boolean bulked = false;
        private String statusMessage = null;
        private String exception = null;

        private Builder() { }

        public Builder code(final HttpResponseStatus code) {
            this.code = code;
            return this;
        }

        public Builder statusMessage(final String message) {
            this.statusMessage = message;
            return this;
        }

        public Builder exception(final String exception) {
            this.exception = exception;
            return this;
        }

        public Builder result(final List<Object> result) {
            this.result = result;
            return this;
        }

        public Builder bulked(final boolean bulked) {
            this.bulked = bulked;
            return this;
        }

        public ResponseMessage create() {
            final ResponseResult responseResult = new ResponseResult(result, bulked);
            // skip null values
            if (code == null && statusMessage == null) {
                return new ResponseMessage(null, responseResult);
            }
            final ResponseStatus responseStatus = new ResponseStatus(code, statusMessage, exception);
            return new ResponseMessage(responseStatus, responseResult);
        }
    }
}
